wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WorkspaceHandler.java

```diff
@@ -0,0 +1,353 @@
+
+      Workspace newWorkspace = Workspace.builder(detail.getWorkspace())
+                                 .setDirectives(executionRequest.getDirectives())
+                                 .setUpdatedTimeMillis(System.currentTimeMillis()).build();
+      store.saveWorkspace(wsId, new WorkspaceDetail(newWorkspace, detail.getSample()));
+      responder.sendJson(response);
+    });
+  }
+
+  /**
+   * Retrieve the directives available in the namespace
+   */
+  @GET
+  @Path("v2/contexts/{context}/directives")
+  public void getDirectives(HttpServiceRequest request, HttpServiceResponder responder,
+                            @PathParam("context") String namespace) {
+    respond(responder, namespace, ns -> {
+      // CDAP-15397 - reload must be called before it can be safely used
+      composite.reload(namespace);
+      List<DirectiveUsage> directives = new ArrayList<>();
+      for (DirectiveInfo directive : composite.list(namespace)) {
+        DirectiveUsage directiveUsage = new DirectiveUsage(directive.name(), directive.usage(),
+                                                           directive.description(), directive.scope().name(),
+                                                           directive.definition(), directive.categories());
+        directives.add(directiveUsage);
+      }
+      responder.sendJson(new ServiceResponse<>(directives));
+    });
+  }
+
+  /**
+   * Get the specification for the workspace
+   */
+  @GET
+  @Path("v2/contexts/{context}/workspaces/{id}/specification")
+  public void specification(HttpServiceRequest request, HttpServiceResponder responder,
+                            @PathParam("context") String namespace,
+                            @PathParam("id") String workspaceId) {
+    respond(responder, namespace, ns -> {
+      if (ns.getName().equalsIgnoreCase(NamespaceId.SYSTEM.getNamespace())) {
+        throw new BadRequestException("Getting specification in system namespace is currently not supported");
+      }
+
+      WorkspaceId wsId = new WorkspaceId(ns, workspaceId);
+      WorkspaceDetail detail = store.getWorkspaceDetail(wsId);
+      List<String> directives = detail.getWorkspace().getDirectives();
+      List<Row> result = executeDirectives(ns.getName(), directives, detail.getSample());
+
+      SchemaConverter schemaConvertor = new SchemaConverter();
+      Schema schema = schemaConvertor.toSchema("record", createUberRecord(result));
+      Map<String, String> properties = ImmutableMap.of("directives", String.join("\n", directives));
+
+      Set<StageSpec> srcSpecs = getSourceSpecs(detail, directives);
+
+      responder.sendString(GSON.toJson(new WorkspaceSpec(
+        srcSpecs, new StageSpec(
+          schema, new Plugin("Wrangler", "transform", properties,
+                             // TODO: find a better way to get the scope and the artifact name
+                             new Artifact("wrangler-transform", ProjectInfo.getVersion(), "system"))))));
```

Comment:
hardcoding this isn't really ideal.

Looks like ArtifactManager.list() is already called (through DirectiveRegistry's reload() method when executing directives). Could we also keep track of the most recent Wrangler when that happens and use it here?

Reply:
Moved the logic to UserDirectiveRegistry to get the wrangler artifact. I keep `ProjectInfo` since we have an /info endpoint which returns the version information, previously it reread the file every time the method is called, now it is in ProjectInfo so the file won't be read each time the endpoint is called